1 collector json

KanID92 opened the pull request.

KanID92 committed:
- v. 0.1;

KanID92 committed:
- v. 0.2;

KanID92 committed:
- v. 0.3;

KanID92 committed:
- v. 0.4;

KanID92 committed:
- v. 0.5;

KanID92 committed:
- v. 0.6;

KanID92 committed:
Sprint 19.

KanID92 committed:
Sprint 19:
-fix: class GeneralAvroSerializer: new SpecificDatumWriter<>(data.getSchema());

kesch9 commented on telemetry/collector/srs/main/java/ru/yandex/practicum/telemetry/collector/service/handler/hub/BaseHubHandler.java:
```diff
@@ -0,0 +1,87 @@
+        properties.put("bootstrap.servers", "localhost:9092");
+        properties.put("client.id", "telemetry.collector");
+        properties.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
+        properties.put("value.serializer", "kafka.serializer.GeneralAvroSerializer");
```
Обязательно
Настройки подключения будут одни и те же к кафки, нужно вынести в отдельный конфигурационный файл

KanID92 replied:
Откорректировано. Не получилось создать конфигурацию с помощью вложенного статического класса, поэтому вложено упрощенно - сразу в поля.

kesch9 commented on telemetry/collector/srs/main/java/ru/yandex/practicum/telemetry/collector/service/handler/hub/DeviceAddedHandler.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Override
+    public void handle(HubEvent hubEvent) {
```
Обязательно
Метод handle один и тот же метод для всех наследников, вынесите его в базовый класс и воспользуйтесь Generic (ProducerRecord<String, T>), а вот метод toAvro будет у каждого свой

KanID92 replied:
Откорректировано.

kesch9 commented on telemetry/collector/srs/main/java/ru/yandex/practicum/telemetry/collector/service/handler/hub/DeviceAddedHandler.java:
```diff
@@ -0,0 +1,39 @@
+    public void handle(HubEvent hubEvent) {
+        ProducerRecord<String, DeviceAddedEventAvro> record = new ProducerRecord<>(
+                topic, toAvro(hubEvent));
+        try(KafkaProducer<String, DeviceAddedEventAvro> producer = new KafkaProducer<>(properties)) {
```
Обязательно
В рамках ТЗ данные нужно записывать в два топика. При этом для записи в оба топика должен использоваться один экземпляр продюсера. Продюсеру в целом без разницы в какой топик отправлять данные. Конкретный топик указывается в экземпляре ProducerRecord с которой уже работает продюсер. Из этого следует, что мы можем использовать один и тот же продюсер для отправки в разные топики. Т.е. producer должен быть одним bean для всех и не создавать каждый раз заново его

KanID92 replied:
Откорректировано.

kesch9 commented on telemetry/collector/srs/main/java/ru/yandex/practicum/telemetry/collector/service/handler/hub/DeviceAddedHandler.java:
```diff
@@ -0,0 +1,39 @@
+        ProducerRecord<String, DeviceAddedEventAvro> record = new ProducerRecord<>(
+                topic, toAvro(hubEvent));
+        try(KafkaProducer<String, DeviceAddedEventAvro> producer = new KafkaProducer<>(properties)) {
+            producer.send(record);
```
Обязательно
При отправке сообщений нужно указывать ключ. Ключом сообщений в оба топика будет идентификатор хаба. Таким образом все события относящиеся к одному и тому же хабу (а значит к одному и тому же умному дому/квартире) будут попадать в одну партицию, а это значит что консьюмер гарантировано будет получать их в правильной очередности (например при параллелизации чтения и обработки топика группой консьюмеров)
Также при создании экземпляра ProducerRecord нужно указывать timestamp записи, в качестве которого брать таймстемп события. Так как мы хотим, чтобы события в партиции были упорядочены по таймстампу когда событие произошло, а не когда данные долетели до брокера кафки
При закрытии продюсера нужно вызвать метод flush, чтобы он успел отправить все данные которые аккумулировались в буфере, а также дать небольшой таймаут на то, чтобы продюсер успел корректно завершить все сетевые соединения и закончить свою работу

KanID92 replied:
Откорректировано.

kesch9 reviewed (commented):
Нужно переделать запись в кафку

KanID92 committed:
Sprint 19. Review fix.
Fix:
- add configuration from application.yaml;
- fix creating of ProducerRecord objects;
- method "handle" move to base handle class;
- add KafkaEventProducer, KafkaConfig classes;

KanID92 committed:
Sprint 19. Review fix.
Fix:
- add sleep for 3 sec for sendRecord method;

kesch9 reviewed (commented):
Approve
